### PR TITLE
Feat: 13 develop tiling processing

### LINKED ISSUE #13: Log recall by class on Yolo models
Image

### PR DESCRIPTION
**Pull request by SoniaGrh**
## Description
This PR introduces a pipeline for dataset tiling using Picsellia. The pipeline extracts, validates, processes, and uploads datasets.

## Issues
This PR resolves #13 of Client Projects Board.

## What's new?
- Added the tiler_processing_pipeline for dataset tiling.
- Integrated steps for dataset extraction, validation, processing, and uploading.

### CODE REVIEW

File: src/models/steps/data_validation/common/object_detection_dataset_context_validator.py [hunk 1]
```diff
@@ -62,16 +62,30 @@ def _validate_bounding_box_coordinates(self, coco_file_annotation: Annotation):
...
-            [
-                x >= 0,
-                y >= 0,
-                width > 0,
-                height > 0,
-                x + width <= coco_file_image.width,
-                y + height <= coco_file_image.height,
-            ]
-        )
+        if x < 0:
```
Comment (picsalex) [inline src/models/steps/data_validation/common/object_detection_dataset_context_validator.py:65]: Instead:
- Count each error type + keep track of the annotations to fix
- Print the Warnings (⚠️)
- Add an extra parameter to the experiment, e.g. `fix_annotation`
- If true: fix, else raise ValueError("...")

File: src/models/steps/data_validation/processing/processing_tiler_data_validator.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            ValueError: If the datalake is not valid.
+        """
+        datalakes_name = [datalake.name for datalake in self.client.list_datalakes()]
+        if self.datalake not in datalakes_name:
+            raise ValueError(
+                f"Datalake {self.datalake} is not valid, available datalakes are {datalakes_name}"
+            )
+
+    def validate(self) -> None:
+        self._validate_tile_size()
```
Comment (picsalex) [inline src/models/steps/data_validation/processing/processing_tiler_data_validator.py:41]: Add `self._validate_ratios()`

File: src/models/steps/processing/dataset_version_creation/dataset_context_uploader.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+from src.models.steps.processing.dataset_version_creation.data_uploader import (
+    DataUploader,
+)
+from src.models.dataset.common.dataset_context import DatasetContext
+
+logger = logging.getLogger("picsellia")
+
+
+class DatasetContextUploader(DataUploader):
```
Comment (picsalex) [inline src/models/steps/processing/dataset_version_creation/dataset_context_uploader.py:16]: Split into two files: `SegmentationDatasetContextUploader` and `ObjectDetectionDatasetContextUploader`.

File: src/models/steps/processing/dataset_version_creation/tiler_processing.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+)
+from shapely.geometry import box, MultiPolygon, GeometryCollection, Polygon
+from shapely.validation import make_valid
+
+from src.models.dataset.processing.processing_dataset_collection import (
+    ProcessingDatasetCollection,
+)
+
+
+def custom_get_sliced_coco_annotation(self, slice_bbox: List[int]):
```
Comment (picsalex) [inline src/models/steps/processing/dataset_version_creation/tiler_processing.py:33]: add docstring

File: src/models/steps/processing/dataset_version_creation/tiler_processing.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+                polygons = [
+                    geom
+                    for geom in valid.geoms
+                    if isinstance(geom, (Polygon, MultiPolygon))
+                ]
+                if polygons:
+                    valid = MultiPolygon(polygons)
+                else:
+                    raise ValueError(
+                        "Aucun polygone valide trouvé dans la GeometryCollection"
```
Comment (picsalex) [inline src/models/steps/processing/dataset_version_creation/tiler_processing.py:49]: in english

File: src/models/steps/processing/dataset_version_creation/tiler_processing.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+                                image_dir: str
+                                    Directory of the sliced image exports.
+                                original_image_size: list of int
+                                    Size of the unsliced original image in [height, width]
+        num_total_invalid_segmentation: int
+            Number of invalid segmentation annotations.
+    """
+
+    # define verboseprint
+    verboselog = logger.info if verbose else lambda *a, **k: None
```
Comment (picsalex) [inline src/models/steps/processing/dataset_version_creation/tiler_processing.py:117]: to remove

File: src/models/steps/processing/dataset_version_creation/tiler_processing.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    if output_file_name and output_dir:
+        conc_exec = concurrent.futures.ThreadPoolExecutor(max_workers=MAX_WORKERS)
+        conc_exec.map(
+            _export_single_slice,
+            sliced_image_result.images,
+            [output_dir] * len(sliced_image_result),
+            sliced_image_result.filenames,
+        )
+
+    verboselog(
```
Comment (picsalex) [inline src/models/steps/processing/dataset_version_creation/tiler_processing.py:226]: Replace with a meaningful `logger.info`

**Comment by picsalex** [comment]
Add the missing `__init__.py` inside `src/models/parameters/common/paddle_ocr`.

File: src/pipelines/dataset_tiler/requirements.in [hunk 8]
```diff
@@ -0,0 +1,6 @@
+imagecodecs
```
Comment (picsalex) [inline src/pipelines/dataset_tiler/requirements.in:1]: Fix the dependencies versions

File: src/steps/data_validation/processing/processing_tiler_data_validator.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+        case InferenceType.SEGMENTATION:
+            object_detection_dataset_validator = ObjectDetectionDatasetContextValidator(
+                dataset_context=dataset_context
+            )
+            segmentation_dataset_validator = SegmentationDatasetContextValidator(
+                dataset_context=dataset_context
+            )
+
+            # SAHI requires bot the bounding boxes and the polygons to be valid
```
Comment (picsalex) [inline src/steps/data_validation/processing/processing_tiler_data_validator.py:49]: typo

**Review by picsalex — CHANGES_REQUESTED** [review]

File: src/models/dataset/common/dataset_context.py [hunk 10]
```diff
@@ -102,7 +101,7 @@ def download_assets(self) -> None:
         if self.multi_asset:
             self.multi_asset.download(target_path=self.image_dir, use_id=self.use_id)
         else:
-            raise ValueError("No assets found in the dataset")
+            print(f"No assets found for dataset {self.dataset_name}!")
```
Comment (picsalex) [inline src/models/dataset/common/dataset_context.py:104]: Do we allow that the dataset is empty?

File: src/models/parameters/training/ultralytics/ultralytics_augmentation_parameters.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+            expected_type=float,
+            default=0.0,
+            range_value=(0.0, 1.0),
+        )
+        self.auto_augment = self.extract_parameter(
+            keys=["auto_augment"], expected_type=str, default="randaugment"
+        )
+        self.erasing = self.extract_parameter(
+            keys=["erasing"], expected_type=float, default=0.4, range_value=(0.0, 1.0)
+        )
```
Comment (picsalex) [inline src/models/parameters/training/ultralytics/ultralytics_augmentation_parameters.py:72]: Aren't we missing the parameter `crop_function`?

Comment (SoniaGrh) [reply src/models/parameters/training/ultralytics/ultralytics_augmentation_parameters.py:72]: J'ai juste déplacé le fichier dans un dossier. Ce sera dans la PR yolov8 classif que je vais compléter ces choses là.

File: src/models/parameters/training/ultralytics/ultralytics_hyper_parameters.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        self.use_cache = self.extract_parameter(
+            keys=["cache", "use_cache"],
+            expected_type=bool,
+            default=False,
+        )
+        self.save_period = self.extract_parameter(
+            keys=["save_period"],
+            expected_type=int,
+            default=-1,
+        )
```
Comment (picsalex) [inline src/models/parameters/training/ultralytics/ultralytics_hyper_parameters.py:22]: I think we should also add the following parameters:
- patience
- optimizer
- cos_lr
- close_mosaic (set to 0 by default)
- amp
- warmup_epoch
- warmup_momentum
- warmup_bias_lr
- overlap_mask
- mask_ratio

WDYT ?

File: src/models/steps/processing/dataset_version_creation/tiler_processing.py [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+                    geom
+                    for geom in valid.geoms
+                    if isinstance(geom, (Polygon, MultiPolygon))
+                ]
+                if polygons:
+                    valid = MultiPolygon(polygons)
+                else:
+                    raise ValueError("No valid polygon found in the GeometryCollection")
+            else:
+                valid = MultiPolygon([valid])
```
Comment (picsalex) [inline src/models/steps/processing/dataset_version_creation/tiler_processing.py:74]: i think we could use more meaningful variable names here (samp/geom/valid, etc)

**Review by picsalex — APPROVED** [review]

**Pull request merged**